src/node_url_pattern.cc

```diff
@@ -0,0 +1,571 @@
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, args.This());
+  if (args.Length() == 0) {  // input, [baseURL]
+    return args.GetReturnValue().SetNull();
+  }
+  auto env = Environment::GetCurrent(args);
+
+  ada::url_pattern_input input;
+  std::optional<std::string> baseURL{};
+  if (args[0]->IsString()) {
+    Utf8Value input_value(env->isolate(), args[0].As<String>());
+    auto url =
+        ada::parse<ada::url_aggregator>(input_value.ToStringView(), nullptr);
+    if (!url) {
+      env->ThrowTypeError("Failed to parse URLPattern");
+      return;
+    }
+    input = std::move(*url);
+  } else if (args[0]->IsObject()) {
+    input = URLPatternInit::FromJsObject(env, args[0].As<Object>());
+  } else {
+    THROW_ERR_INVALID_ARG_TYPE(
+        env, "URLPattern input needs to be a string or an object");
+    return;
+  }
+
+  if (args.Length() > 1 && args[1]->IsString()) {
+    Utf8Value base_url_value(env->isolate(), args[1].As<String>());
+    baseURL = base_url_value.ToStringView();
+  }
+
+  Local<Value> result;
+  std::optional<std::string_view> baseURL_opt =
+      baseURL ? std::optional<std::string_view>(*baseURL) : std::nullopt;
+  if (!url_pattern->Exec(env, std::move(input), baseURL_opt).ToLocal(&result)) {
+    args.GetReturnValue().Set(Null(args.GetIsolate()));
+    return;
+  }
+  args.GetReturnValue().Set(result);
+}
+
+void URLPattern::Test(const FunctionCallbackInfo<Value>& args) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, args.This());
+  if (args.Length() == 0) {  // input, [baseURL]
+    return args.GetReturnValue().Set(false);
+  }
+  auto env = Environment::GetCurrent(args);
+
+  ada::url_pattern_input input;
+  std::optional<std::string> baseURL{};
+  if (args[0]->IsString()) {
+    Utf8Value input_value(env->isolate(), args[0].As<String>());
+    auto url =
+        ada::parse<ada::url_aggregator>(input_value.ToStringView(), nullptr);
+    if (!url) {
+      env->ThrowTypeError("Failed to parse URLPattern");
+      return;
+    }
+    input = std::move(*url);
+  } else if (args[0]->IsObject()) {
+    input = URLPatternInit::FromJsObject(env, args[0].As<Object>());
+  } else {
+    THROW_ERR_INVALID_ARG_TYPE(
+        env, "URLPattern input needs to be a string or an object");
+    return;
+  }
+
+  if (args.Length() > 1 && args[1]->IsString()) {
+    Utf8Value base_url_value(env->isolate(), args[1].As<String>());
+    baseURL = base_url_value.ToStringView();
+  }
+
+  std::optional<std::string_view> baseURL_opt =
+      baseURL ? std::optional<std::string_view>(*baseURL) : std::nullopt;
+  args.GetReturnValue().Set(
+      url_pattern->Test(env, std::move(input), baseURL_opt));
+}
+
+void URLPattern::Protocol(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Protocol().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
```

Comment:
```suggestion
  if (url_pattern->Protocol().ToLocal(&result)) {
    info.GetReturnValue().Set(result);
  }
```

a bit more compact to invert the checks on these.